lanl/BEE
# General logging implementation: restructure

This PR implements logging for BEE, and converts all bee component logging mechanisms to the new logging style based on the #274. This work was only possible after the PyPA code restructure, which is why I created a new PR.

I will demonstrate how the system works on Wednesday, but feel free to test at your convenience. Expected behavior is this:

1. All log calls in cli.py itself stream to STDIO (this includes gdb and restd, as described in #276.
2. All logs associated with cli.py are stored in `{bee_workdir}/logs/beeflow.log`.
3. All logs associated with specific BEE components and sub-components (flask, werkzeug) stream to `{bee_workdir}/logs/{component_name}.log` until the process is killed or crashes. 

Notes:

- Werkzeug appears to have an error associated with accepting external logging handlers. See this issue I filed: https://github.com/pallets/werkzeug/issues/2002. I pestered them in their community chat and got no responses, so assuming it's a bug for now.
- Because we are not using dedicated launch scripts for gdb and restd, the logging is a little clunky. If we make gdb and restd launch in a subprocess script like the other stuff, then we only have to use one logger for all bee components. As it stands, we have to use 3 [1]. See issue #276.
- I changed all components logs in one way or another. You shouldn't need to change anything, but it might be good to verify that your scripts are reporting in a way that makes sense to you. Let me know if I can help.
- I tried to turn all of our "print" statements into "log" calls. If I missed some of yours, please change them or these messages will be lost to oblivion.

# How to test
1. Check out this branch on Fog
2. Clean up any previous logs from `{bee_workdir}/logs`
2. `poetry install` from project root
3. Be sure poetry environment is activated
4. `cd src/` (this necessity will go away with #275)
5. `beeflow --sleep-time 45`
6. Connect client
7. Load workflow in it's new path relative to src: `beeflow/data/cwl/cf.cwl`
8. Run workflow
9. Observe CLAMR movie in `$HOME`. Observe logs from all components relative to `{bee_workdir}/logs`.

### References
[1]Logging will be less clunky and more efficient if GDB and restd get their own launch scripts.
https://github.com/lanl/BEE_Private/blob/41f2ccc60388da483a49f52d18b1677ba9a98f3f/src/beeflow/cli.py#L26-L28
